test/unit/bitmap_test.c

```diff
@@ -0,0 +1,222 @@
+ */ 
+boolean test_copy(heap h, bitmap b) { 
+    bitmap b_cpy = allocate_bitmap(h, h, 4096);
+    for (int i = 0; i < 20; i++) {
+        bitmap_set(b_cpy, rand(), 1);
+        bitmap_set(b, rand(), 1);
+    }
+    bitmap_copy(b, b_cpy);
+    // tests bits match in original bitmap
+    bitmap_foreach_set(b, j) {
+        // implicit test for bitmap_foreach_set
+        if (!bitmap_get(b, j)) {
+            msg_err("!!! foreach_set failed for bitmap\n");
+            deallocate_bitmap(b_cpy);
+            return false;
+        }
+        if (!bitmap_get(b_cpy, j)) {
+            msg_err("!!! cloning failed for bitmap\n");
+            deallocate_bitmap(b_cpy);
+            return false;
+        }
+    }
+    // tests bits match in copy of bitmap
+    bitmap_foreach_set(b_cpy, j) {
+        // implicit test for bitmap_foreach_set
+        if (!bitmap_get(b_cpy, j)) {
+            msg_err("!!! foreach_set failed for bitmap\n");
+            deallocate_bitmap(b_cpy);
+            return false;
+        }
+        if (!bitmap_get(b, j)) {
+            msg_err("!!! cloning failed for bitmap\n");
+            deallocate_bitmap(b_cpy);
+            return false;
+        }
+    }
+    deallocate_bitmap(b_cpy);
+    return true;
+}
+
+/**
+ *  Tests getting and setting bits of a bitmap using 
+ *  bitmap_get and bitmap_set functions.
+ */ 
+boolean test_set_and_get(bitmap b) {
+    u64 i = rand();
+    bitmap_set(b, i, 1);
+    if (!bitmap_get(b, i)) {
+        msg_err("!!! set and get failed for bitmap\n");
+        return false;
+    }
+    return true;
+}
+
+/**
+ *  Tests wrapping and unwrapping of bitmap using 
+ *  bitmap_wrap and bitmap_unwrap functions.
+ */ 
+boolean test_wrap(heap h) { 
+    u64 map = rand(); 
+    map = (map << 32) | rand();
+    bitmap b = bitmap_wrap(h, &map, 64);
+    for (int i = 0; i < 64; i++) {
+        if (((map & (1 << i)) && !bitmap_get(b, i)) || (!(map & (1 << i)) && bitmap_get(b, i))) {
+            msg_err("!!! wrap failed for bitmap at bit %d | map: %d bitmap: %d\n", i, (map & (1 << i)), bitmap_get(b, i));
+            bitmap_unwrap(b);
+            return false;
+        }
+    }
+    bitmap_unwrap(b);
+    return true;
+}
+
+/**
+ *  Tests bitmap alloc and dealloc of bitmap using 
+ *  bitmap_alloc_within_range and bitmap_dealloc functions.
+ */ 
+boolean test_bitmap_alloc(bitmap b, u64 start, u64 end) {
+    // generate random number between 0 and (end-start)
+    u64 nbits = rand() % (end - start + 1);
+    u64 first = bitmap_alloc_within_range(b, nbits, start, end);
+    if ((first == INVALID_PHYSICAL) || !bitmap_dealloc(b, first, nbits)) {
+        msg_err("!!! alloc range failed for bitmap\n");
+        return false;
+    }
+    return true;
+}
+
+/**
+ *  Tests bitmap range check and set of bitmap using
+ *  bitmap_range_check_and_set
+ */
+boolean test_range_check_set(bitmap b) {
+    u64 start = rand() % 4096;
+    u64 nbits = rand() % (4097 - start);
+    boolean validate = rand() & 1;
```

Comment:
Since the bitmap at this point may have some random bits set (because it has been used in the functions called prior to `(test_range_check_set)`), you cannot be sure that the validation step in bitmap_range_check_and_set() will pass. So depending on what random values are generated, this test may pass or fail. To avoid this possibility, I would just set `validate` to false.